examples/io_test/io_bench.f90

```diff
@@ -40,11 +42,20 @@ program io_bench
 
    call decomp_2d_testing_log()
 
+   call decomp_2d_io_init()
+   call decomp_2d_io_register_var3d("io.dat", 1, real_type)
+
    call alloc_x(u1, .true.)
    call random_number(u1)
 
    t1 = MPI_WTIME()
-   call decomp_2d_write_one(1, u1, 'io.dat')
+#ifdef ADIOS2
+   call io%open_start("./", decomp_2d_write_mode)
+#endif
+   call decomp_2d_write_one(1, u1, 'io.dat', opt_dirname="./", opt_io=io)
+#ifdef ADIOS2
+   call io%end_close
+#endif
```

Comment:
Could we handle `#ifdef ADIOS2` inside the `io%open_start/close` calls?

Reply:
No, as with previous design, this allows creating a file per variable with MPIIO when not opened or appending multiple data to a single file when an open file exists. By having an `io` object though, perhaps we can improve on this design? e.g.
`
io%open_start(..., opt_write_multiple=.true.) ! Writes multiple data to one file
io%open_start(..., opt_write_multiple=.false.) ! Writes single variable to file (default?)
decomp_2d_write_one(..., opt_io=io) ! Writes according to options set on io object, opens single file if none open
`

Reply:
The current design for `decomp_2d_write_one` allows to

1. Write only one variable in one file. This is possible if a IO object is provided or if a path is provided.
2. Write several variables in one file. This is possible if a IO object is provided.

I have tried to avoid mandatory IO objects for external codes using MPI... To move the IO open / start / end / close calls inside the subroutine `decomp_2d_write_one`, I think we should have one interface for writing one variable in one file. And another interface for writing several variables in one file ?

Reply:
ah, I thought we wanted to make the IO objects mandatory.

Reply:
We could, the opt_dirname argument could then be removed. And some code dedicated to cases without the IO object argument could be removed. https://github.com/2decomp-fft/2decomp-fft/blob/refactor_io/src/io.f90#L1107

Reply:
As a side remark, I think the lines
```
#ifdef ADIOS2
   call io%open_start("./", decomp_2d_write_mode)
#endif
```
and
```
#ifdef ADIOS2
   call io%end_close
#endif
```
are not mandatory for ADIOS2 builds : line 1107 in io.f90 will create a IO object if none is provided.

Reply:
Fixed in https://github.com/2decomp-fft/2decomp-fft/pull/269/commits/d35d0ecb24a5b8abbb8999c73f113960cdcc0de9